clients/client/src/main/java/org/projectnessie/client/http/ApiHttpRequest.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Copyright (C) 2022 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.client.http;
+
+import java.util.function.Supplier;
+
+/**
+ * This is a helper class for use with an {@link HttpRequest} that will unwrap the specified
+ * API-level exceptions from {@link HttpClientException} thrown during the execution of the {@link
+ * HttpRequest}.
+ *
+ * <p>Currently this class supports up to two distinct API-level exception types, but it can easily
+ * be extended to support more if required.
+ *
+ * <p>The exception types to be unwrapped are specifies as arguments to {@link
+ * HttpRequest#unwrap(Class)} and {@link HttpRequest#unwrap(Class, Class)} calls.
+ *
+ * @param <E1> the first API-level exception that should be unwrapped
+ * @param <E2> the second API-level exception that should be unwrapped
+ */
+public class ApiHttpRequest<E1 extends Throwable, E2 extends Throwable> {
```

Comment:
Does it need to be `public`? (Isn't package-private good enough? Maybe with a wrapping interface - but w/o over-engineering - so only, if it's a simple thing to do)

Reply:
will do another follow-up commit for this

Reply:
I extracted an interface for now - see the latest review commit.

It might be nicer just to fold the exception unwrapping logic into `HttpRequest` itself, but I did not want to touch a lot of v1 code with minor refactoring due to type parameter changes. We could do that in a follow-up PR. WDYT?